modules/webgl/src/adapter/resources/webgl-transform-feedback.ts

```diff
@@ -0,0 +1,253 @@
+
+  _bindBuffer(
+    index: number,
+    buffer: WEBGLBuffer,
+    byteOffset: number = 0,
+    byteLength?: number
+  ): this {
+    const handle = buffer && buffer.handle;
+    if (!handle || byteLength === undefined) {
+      // @ts-expect-error
+      this.gl.bindBufferBase(GL.TRANSFORM_FEEDBACK_BUFFER, index, handle);
+    } else {
+      // @ts-expect-error
+      this.gl.bindBufferRange(GL.TRANSFORM_FEEDBACK_BUFFER, index, handle, byteOffset, byteLength);
+    }
+    return this;
+  }
+
+  // SUBCLASS
+
+  initialize(props?: TransformFeedbackProps): this {
+    this.buffers = {};
+    this.unusedBuffers = {};
+    this.bindOnUse = true;
+
+    // Unbind any currently bound buffers
+    if (!isObjectEmpty(this.buffers)) {
+      this.bind(() => this._unbindBuffers());
+    }
+
+    this.setProps(props);
+    return this;
+  }
+
+  setProps(props: TransformFeedbackProps) {
+    if ('buffers' in props) {
+      this.setBuffers(props.buffers);
+    }
+  }
+
+  setBuffers(buffers = {}) {
+    this.bind(() => {
+      for (const bufferName in buffers) {
+        this.setBuffer(bufferName, buffers[bufferName]);
+      }
+    });
+    return this;
+  }
+
+  setBuffer(locationOrName: string | number, bufferOrRange: WEBGLBuffer) {
+    const location = this._getVaryingIndex(locationOrName);
+    const {buffer, byteLength, byteOffset} = this._getBufferRange(bufferOrRange);
+
+    if (location < 0) {
+      this.unusedBuffers[locationOrName] = buffer;
+      log.warn(`${this.id} unusedBuffers varying buffer ${locationOrName}`)();
+      return this;
+    }
+
+    this.buffers[location] = {buffer, byteLength, byteOffset};
+
+    // Need to avoid chrome bug where buffer that is already bound to a different target
+    // cannot be bound to 'TRANSFORM_FEEDBACK_BUFFER' target.
+    if (!this.bindOnUse) {
+      this._bindBuffer(location, buffer, byteOffset, byteLength);
+    }
+
+    return this;
+  }
+
+  // PRIVATE METHODS
+
+  /** Extract offsets for bindBufferRange */
+  _getBufferRange(
+    bufferOrRange: WEBGLBuffer | {buffer: WEBGLBuffer; byteOffset?: number; byteLength?: number}
+  ): Required<BufferRange> {
+    if (bufferOrRange instanceof WEBGLBuffer) {
+      return {buffer: bufferOrRange, byteOffset: 0, byteLength: bufferOrRange.byteLength};
+    }
+
+    // To use bindBufferRange either offset or size must be specified.
+    const {buffer, byteOffset = 0, byteLength = bufferOrRange.buffer.byteLength} = bufferOrRange;
+    return {buffer, byteOffset, byteLength};
+  }
+
+  _getVaryingInfo(locationOrName: string | number) {
+    throw new Error(`${this.id} getVaryingInfo not implemented`);
+  }
+
+  _getVaryingIndex(locationOrName: string | number) {
+    if (typeof locationOrName === 'number') {
+      return locationOrName;
+    }
+
+    for (const varying of this.layout.varyings) {
+      if (locationOrName === varying.name) {
+        return varying.location;
+      }
+    }
+
+    return -1;
+  }
+
+  bind(funcOrHandle = this.handle) {
```

Comment:
Since these are public I would move them just above the private/protected methods. I like to scan the public API of a class from top down, and stop reading when I get to private methods.

Reply:
Done!